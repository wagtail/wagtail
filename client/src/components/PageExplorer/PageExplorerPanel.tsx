

import React from 'react';

import { STRINGS, MAX_EXPLORER_PAGES } from '../../config/wagtailConfig';

import LoadingSpinner from '../LoadingSpinner/LoadingSpinner';
import Transition, { PUSH, POP } from '../Transition/Transition';
import PageExplorerHeader from './PageExplorerHeader';
import PageExplorerItem from './PageExplorerItem';
import PageCount from './PageCount';
import { State as NodeState, PageState } from './reducers/nodes';

interface PageExplorerPanelProps {
  nodes: NodeState;
  depth: number;
  page: PageState;
  onClose(): void;
  gotoPage(id: number, transition: number): void;
  navigate(url: string): Promise<void>;
}

interface PageExplorerPanelState {
  transition: typeof PUSH | typeof POP;
}

/**
 * The main panel of the page explorer menu, with heading,
 * menu items, and special states.
 */
class PageExplorerPanel extends React.Component<PageExplorerPanelProps, PageExplorerPanelState> {
  constructor(props) {
    super(props);

    this.state = {
      transition: PUSH,
    };

    this.onItemClick = this.onItemClick.bind(this);
    this.onHeaderClick = this.onHeaderClick.bind(this);
  }

  componentWillReceiveProps(newProps) {
    const { depth } = this.props;
    const isPush = newProps.depth > depth;

    this.setState({
      transition: isPush ? PUSH : POP,
    });
  }

  onItemClick(id, e) {
    const { gotoPage } = this.props;

    e.preventDefault();
    e.stopPropagation();

    gotoPage(id, 1);
  }

  onHeaderClick(e) {
    const { page, depth, gotoPage } = this.props;
    const parent = page.meta.parent?.id;

    // Note: Checking depth as well in case the user started deep in the tree
    if (depth > 0 && parent) {
      e.preventDefault();
      e.stopPropagation();

      gotoPage(parent, -1);
    }
  }

  renderChildren() {
    const { page, nodes } = this.props;
    let children;

    if (!page.isFetchingChildren && !page.children.items) {
      children = (
        <div key="empty" className="c-page-explorer__placeholder">
          {STRINGS.NO_RESULTS}
        </div>
      );
    } else {
      children = (
        <div key="children">
          {page.children.items.map((id) => (
            <PageExplorerItem
              key={id}
              item={nodes[id]}
              onClick={this.onItemClick.bind(null, id)}
              navigate={this.props.navigate}
            />
          ))}
        </div>
      );
    }

    return (
      <div className="c-page-explorer__drawer">
        {children}
        {page.isFetchingChildren || page.isFetchingTranslations ? (
          <div key="fetching" className="c-page-explorer__placeholder">
            <LoadingSpinner />
          </div>
        ) : null}
        {page.isError ? (
          <div key="error" className="c-page-explorer__placeholder">
            {STRINGS.SERVER_ERROR}
          </div>
        ) : null}
      </div>
    );
  }

  render() {
    const { page, depth, gotoPage } = this.props;
    const { transition } = this.state;

    return (
      <Transition name={transition} className="c-page-explorer" component="nav" label={STRINGS.PAGE_EXPLORER}>
        <div key={depth} className="c-transition-group">
          <PageExplorerHeader
            depth={depth}
            page={page}
            onClick={this.onHeaderClick}
            gotoPage={gotoPage}
            navigate={this.props.navigate}
          />

          {this.renderChildren()}

          {page.isError || page.children.items && page.children.count > MAX_EXPLORER_PAGES ? (
            <PageCount page={page} />
          ) : null}
        </div>
      </Transition>
    );
  }
}

export default PageExplorerPanel;
