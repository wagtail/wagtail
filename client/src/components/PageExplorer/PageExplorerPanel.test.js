import React from 'react';
import { shallow } from 'enzyme';
import PageExplorerPanel from './PageExplorerPanel';

const mockProps = {
  page: {
    children: {
      items: [],
    },
    meta: {
      parent: null,
    },
  },
  depth: 1,
  onClose: jest.fn(),
  gotoPage: jest.fn(),
  nodes: {},
};

describe('PageExplorerPanel', () => {
  describe('general rendering', () => {
    beforeEach(() => {
      document.body.innerHTML = '<div></div>';
    });

    it('exists', () => {
      expect(PageExplorerPanel).toBeDefined();
    });

    it('renders', () => {
      expect(shallow(<PageExplorerPanel {...mockProps} />)).toMatchSnapshot();
    });

    it('#isFetching', () => {
      expect(
        shallow(
          <PageExplorerPanel
            {...mockProps}
            page={{ isFetching: true, ...mockProps.page }}
          />,
        ),
      ).toMatchSnapshot();
    });

    it('#isError', () => {
      expect(
        shallow(
          <PageExplorerPanel
            {...mockProps}
            page={{ isError: true, ...mockProps.page }}
          />,
        ),
      ).toMatchSnapshot();
    });

    it('no children', () => {
      expect(
        shallow(<PageExplorerPanel {...mockProps} page={{ children: {} }} />),
      ).toMatchSnapshot();
    });

    it('#items', () => {
      expect(
        shallow(
          <PageExplorerPanel
            {...mockProps}
            page={{ children: { items: [1, 2] } }}
            nodes={{
              1: {
                id: 1,
                admin_display_title: 'Test',
                meta: { status: {}, type: 'test' },
              },
              2: {
                id: 2,
                admin_display_title: 'Foo',
                meta: { status: {}, type: 'foo' },
              },
            }}
          />,
        ),
      ).toMatchSnapshot();
    });
  });

  describe('onHeaderClick', () => {
    beforeEach(() => {
      mockProps.gotoPage.mockReset();
    });

    it('calls gotoPage', () => {
      shallow(
        <PageExplorerPanel
          {...mockProps}
          depth={2}
          page={{ children: { items: [] }, meta: { parent: { id: 1 } } }}
        />,
      )
        .find('PageExplorerHeader')
        .prop('onClick')({
        preventDefault() {},
        stopPropagation() {},
      });

      expect(mockProps.gotoPage).toHaveBeenCalled();
    });

    it('does not call gotoPage for first page', () => {
      shallow(
        <PageExplorerPanel
          {...mockProps}
          depth={0}
          page={{ children: { items: [] }, meta: { parent: { id: 1 } } }}
        />,
      )
        .find('PageExplorerHeader')
        .prop('onClick')({
        preventDefault() {},
        stopPropagation() {},
      });

      expect(mockProps.gotoPage).not.toHaveBeenCalled();
    });
  });

  describe('onItemClick', () => {
    beforeEach(() => {
      mockProps.gotoPage.mockReset();
    });

    it('calls gotoPage', () => {
      shallow(
        <PageExplorerPanel
          {...mockProps}
          path={[1]}
          page={{ children: { items: [1] } }}
          nodes={{
            1: {
              id: 1,
              admin_display_title: 'Test',
              meta: { status: {}, type: 'test' },
            },
          }}
        />,
      )
        .find('PageExplorerItem')
        .prop('onClick')({
        preventDefault() {},
        stopPropagation() {},
      });

      expect(mockProps.gotoPage).toHaveBeenCalled();
    });
  });

  describe('hooks', () => {
    it('componentWillReceiveProps push', () => {
      const wrapper = shallow(<PageExplorerPanel {...mockProps} />);
      expect(wrapper.setProps({ depth: 2 }).state('transition')).toBe('push');
    });

    it('componentWillReceiveProps pop', () => {
      const wrapper = shallow(<PageExplorerPanel {...mockProps} />);
      expect(wrapper.setProps({ depth: 0 }).state('transition')).toBe('pop');
    });
  });
});
