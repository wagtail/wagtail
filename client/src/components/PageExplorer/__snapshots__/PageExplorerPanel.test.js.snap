// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PageExplorerPanel general rendering #isError 1`] = `
<Transition
  className="c-page-explorer"
  component="nav"
  duration={210}
  label="Page explorer"
  name="push"
>
  <div
    className="c-transition-group"
    key="1"
  >
    <PageExplorerHeader
      depth={1}
      gotoPage={[MockFunction]}
      onClick={[Function]}
      page={
        Object {
          "children": Object {
            "items": Array [],
          },
          "isError": true,
          "meta": Object {
            "parent": null,
          },
        }
      }
    />
    <div
      className="c-page-explorer__drawer"
    >
      <div
        key="children"
      />
      <div
        className="c-page-explorer__placeholder"
        key="error"
      >
        Server Error
      </div>
    </div>
    <PageCount
      page={
        Object {
          "children": Object {
            "items": Array [],
          },
          "isError": true,
          "meta": Object {
            "parent": null,
          },
        }
      }
    />
  </div>
</Transition>
`;

exports[`PageExplorerPanel general rendering #isFetching 1`] = `
<Transition
  className="c-page-explorer"
  component="nav"
  duration={210}
  label="Page explorer"
  name="push"
>
  <div
    className="c-transition-group"
    key="1"
  >
    <PageExplorerHeader
      depth={1}
      gotoPage={[MockFunction]}
      onClick={[Function]}
      page={
        Object {
          "children": Object {
            "items": Array [],
          },
          "isFetching": true,
          "meta": Object {
            "parent": null,
          },
        }
      }
    />
    <div
      className="c-page-explorer__drawer"
    >
      <div
        key="children"
      />
    </div>
  </div>
</Transition>
`;

exports[`PageExplorerPanel general rendering #items 1`] = `
<Transition
  className="c-page-explorer"
  component="nav"
  duration={210}
  label="Page explorer"
  name="push"
>
  <div
    className="c-transition-group"
    key="1"
  >
    <PageExplorerHeader
      depth={1}
      gotoPage={[MockFunction]}
      onClick={[Function]}
      page={
        Object {
          "children": Object {
            "items": Array [
              1,
              2,
            ],
          },
        }
      }
    />
    <div
      className="c-page-explorer__drawer"
    >
      <div
        key="children"
      >
        <PageExplorerItem
          item={
            Object {
              "admin_display_title": "Test",
              "id": 1,
              "meta": Object {
                "status": Object {},
                "type": "test",
              },
            }
          }
          key="1"
          onClick={[Function]}
        />
        <PageExplorerItem
          item={
            Object {
              "admin_display_title": "Foo",
              "id": 2,
              "meta": Object {
                "status": Object {},
                "type": "foo",
              },
            }
          }
          key="2"
          onClick={[Function]}
        />
      </div>
    </div>
  </div>
</Transition>
`;

exports[`PageExplorerPanel general rendering no children 1`] = `
<Transition
  className="c-page-explorer"
  component="nav"
  duration={210}
  label="Page explorer"
  name="push"
>
  <div
    className="c-transition-group"
    key="1"
  >
    <PageExplorerHeader
      depth={1}
      gotoPage={[MockFunction]}
      onClick={[Function]}
      page={
        Object {
          "children": Object {},
        }
      }
    />
    <div
      className="c-page-explorer__drawer"
    >
      <div
        className="c-page-explorer__placeholder"
        key="empty"
      >
        No results
      </div>
    </div>
  </div>
</Transition>
`;

exports[`PageExplorerPanel general rendering renders 1`] = `
<Transition
  className="c-page-explorer"
  component="nav"
  duration={210}
  label="Page explorer"
  name="push"
>
  <div
    className="c-transition-group"
    key="1"
  >
    <PageExplorerHeader
      depth={1}
      gotoPage={[MockFunction]}
      onClick={[Function]}
      page={
        Object {
          "children": Object {
            "items": Array [],
          },
          "meta": Object {
            "parent": null,
          },
        }
      }
    />
    <div
      className="c-page-explorer__drawer"
    >
      <div
        key="children"
      />
    </div>
  </div>
</Transition>
`;
